examples/chef/common/clusters/switch/SwitchManager.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "SwitchEventHandler.h"
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app/clusters/switch-server/switch-server.h>
+#include <app/server/Server.h>
+#include <app/util/att-storage.h>
+#include <app/util/attribute-storage.h>
+#include <platform/PlatformManager.h>
+
+#include "chef-descriptor-namespace.h"
+#include "chef-rpc-actions-worker.h"
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Switch;
+using namespace chip::DeviceLayer;
+
+using namespace chip::rpc;
+using namespace chip::app;
+
+class SwitchActionsDelegate : public chip::app::ActionsDelegate
+{
+public:
+    SwitchActionsDelegate(ClusterId clusterId, SwitchEventHandler * eventHandler) :
+        ActionsDelegate(clusterId), mEventHandler(eventHandler){};
+    ~SwitchActionsDelegate() override{};
+
+    void AttributeWriteHandler(chip::EndpointId endpointId, chip::AttributeId attributeId, std::vector<uint32_t> args) override;
+    void EventHandler(chip::EndpointId endpointId, chip::EventId eventId, std::vector<uint32_t> args) override;
+
+private:
+    SwitchEventHandler * mEventHandler;
+};
+
+void SwitchActionsDelegate::AttributeWriteHandler(chip::EndpointId endpointId, chip::AttributeId attributeId,
+                                                  std::vector<uint32_t> args)
+{
+    if (args.empty())
+    {
+        ChipLogError(NotSpecified, "Queue is empty ");
+        return;
+    }
+
+    switch (attributeId)
+    {
+    case Switch::Attributes::NumberOfPositions::Id: {
+        uint8_t data = static_cast<uint8_t>(args[0]);
+        app::Clusters::Switch::Attributes::NumberOfPositions::Set(endpointId, data);
```

Comment:
I don't know what exactly your question is. If you mean that if this is called by `chip-tool`, it won't since the attribute is read-only. So this would only be called by RPC


```
(gdb) where
#0  SwitchActionsDelegate::AttributeWriteHandler(unsigned short, unsigned int, std::vector<unsigned int, std::allocator<unsigned int> >)
    (this=0x5555557e3110, endpointId=1, attributeId=1, args=std::vector of length 1, capacity 1 = {...})
    at /home/erwinpan/matter/erwinpan1/master_0515_genericswitch/examples/chef/common/clusters/switch/SwitchManager.cpp:75
#1  0x000055555557a38e in chip::app::ChefRpcActionsWorker::ProcessActionQueue() (this=0x5555557b9800 <instance>)
    at /home/erwinpan/matter/erwinpan1/master_0515_genericswitch/examples/chef/common/chef-rpc-actions-worker.cpp:103
#2  0x0000555555579f72 in RpcActionsTaskCallback(chip::System::Layer*, void*)
    (systemLayer=0x5555557bc1a0 <chip::DeviceLayer::SystemLayerImpl()::gSystemLayerImpl>, data=0x5555557b9800 <instance>)
    at /home/erwinpan/matter/erwinpan1/master_0515_genericswitch/examples/chef/common/chef-rpc-actions-worker.cpp:51
#3  0x0000555555662f2f in chip::System::TimerData::Callback::Invoke() const (this=0x7fffffffcf20) at ../third_party/connectedhomeip/src/system/SystemTimer.h:60
#4  0x00005555556648f3 in chip::System::TimerPool<chip::System::TimerList::Node>::Invoke(chip::System::TimerList::Node*)
    (this=0x5555557bc7a8 <chip::DeviceLayer::SystemLayerImpl()::gSystemLayerImpl+1544>, timer=0x7fffec000c10) at ../third_party/connectedhomeip/src/system/SystemTimer.h:236
#5  0x0000555555664411 in chip::System::LayerImplSelect::HandleEvents() (this=0x5555557bc1a0 <chip::DeviceLayer::SystemLayerImpl()::gSystemLayerImpl>)
    at ../third_party/connectedhomeip/src/system/SystemLayerImplSelect.cpp:683
#6  0x000055555569dfbe in chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::_RunEventLoop()
    (this=0x5555557bf010 <chip::DeviceLayer::PlatformManagerImpl::sInstance+16>)
    at ../third_party/connectedhomeip/src/include/platform/internal/GenericPlatformManagerImpl_POSIX.ipp:225
#7  0x00005555555c7d10 in chip::DeviceLayer::PlatformManager::RunEventLoop() (this=0x5555557bf000 <chip::DeviceLayer::PlatformManagerImpl::sInstance>)
    at ../third_party/connectedhomeip/src/include/platform/PlatformManager.h:403
#8  0x00005555555ca734 in ChipLinuxAppMainLoop(AppMainLoopImplementation*) (impl=0x0) at ../third_party/connectedhomeip/examples/platform/linux/AppMain.cpp:618
#9  0x000055555557f534 in main(int, char**) (argc=1, argv=0x7fffffffd2b8) at ../main.cpp:52
```

Reply:
It seems like the question is if this is always from the same thread as the chip event loop since setters require you to be in the event loop. The stack trace seems to indicate we are in `RunEventLoop` on a timer invocation.